spring-integration-zmq/src/main/java/org/springframework/integration/zmq/inbound/ZmqMessageDrivenChannelAdapter.java

```diff
@@ -0,0 +1,326 @@
+
+	private volatile ZMQ.Context context;
+
+	private volatile ZMQ.Socket client;
+
+	private volatile ZMQ.Poller poller;
+
+	private volatile ScheduledFuture<?> reconnectFuture;
+
+	private volatile boolean connected;
+
+	private volatile int recoveryInterval = DEFAULT_RECOVERY_INTERVAL;
+
+	private volatile boolean cleanSession;
+
+	private volatile ConsumerStopAction consumerStopAction;
+
+	private ApplicationEventPublisher applicationEventPublisher;
+
+	/**
+	 * Use this constructor for a single url (although it may be overridden
+	 * if the server URI is provided by the {@link DefaultZmqClientFactory}.
+	 * @param url the URL.
+	 * @param clientId The client id.
+	 * @param clientFactory The client factory.
+	 */
+	public ZmqMessageDrivenChannelAdapter(String url, String clientId,
+										ZmqClientFactory clientFactory) {
+		super(url, clientId);
+		this.clientFactory = clientFactory;
+	}
+
+	/**
+	 * Use this constructor if the server URI is provided by the {@link ZmqClientFactory}.
+	 * @param clientId The client id.
+	 * @param clientFactory The client factory.
+	 * @since 5.0.1
+	 */
+	public ZmqMessageDrivenChannelAdapter(String clientId, ZmqClientFactory clientFactory) {
+		super(null, clientId);
+		this.clientFactory = clientFactory;
+	}
+
+	/**
+	 * Use this URL when you don't need additional {@link ZmqClientFactory}.
+	 * @param url The URL.
+	 * @param clientId The client id.
+	 */
+	public ZmqMessageDrivenChannelAdapter(String url, String clientId) {
+		this(url, clientId, new DefaultZmqClientFactory(ZMQ.SUB));
+	}
+
+	/**
+	 * The time (ms) to wait between reconnection attempts.
+	 * Default {@value #DEFAULT_RECOVERY_INTERVAL}.
+	 * @param recoveryInterval the interval.
+	 * @since 5.0.1
+	 */
+	public void setRecoveryInterval(int recoveryInterval) {
+		this.recoveryInterval = recoveryInterval;
+	}
+
+	/**
+	 * @since 5.0.1
+	 */
+	@Override
+	public void setApplicationEventPublisher(ApplicationEventPublisher applicationEventPublisher) {
+		this.applicationEventPublisher = applicationEventPublisher;
+	}
+
+	@Override
+	protected void doStart() {
+		Assert.state(getTaskScheduler() != null, "A 'taskScheduler' is required");
+		super.doStart();
+		try {
+			connectAndSubscribe();
+		}
+		catch (Exception e) {
+			connectionLost(e);
+		}
+	}
+
+	public synchronized void connectionLost(Throwable cause) {
+		logger.error("Exception while connecting and subscribing, retrying", cause);
+		this.connected = false;
+		this.scheduleReconnect();
+		if (this.applicationEventPublisher != null) {
+			this.applicationEventPublisher.publishEvent(new ZmqConnectionFailedEvent(this, cause));
+		}
+	}
+
+	@Override
+	protected void doStop() {
+		cancelReconnect();
+		super.doStop();
+		if (this.client != null) {
+			try {
+				if (this.consumerStopAction.equals(ConsumerStopAction.UNSUBSCRIBE_ALWAYS) ||
+						this.consumerStopAction.equals(ConsumerStopAction.UNSUBSCRIBE_CLEAN) &&
+						this.cleanSession) {
```

Comment:
Do we really need all this logic?
Does it really here work as in MQTT protocol?